### PR TITLE
[ALLUXIO-2560] Improving dependency and plugin management

### PR DESCRIPTION
**Pull request by jsimsa**
Addresses https://alluxio.atlassian.net/browse/ALLUXIO-2560

Dependency versions are specified in the root module using dependency management.

Plugin versions are specified in the root module using plugin management.

All dependencies are sorted as follows:

- dependencies are divided into external, internal, external test, and internal test groups
- within each group dependencies are sorted lexicographically by group ID and artifact ID

Replacement for #4770

### CODE REVIEW

**Comment by jsimsa** [comment]
@aaudiber @apc999 @calvinjia I decided to split #4770 into two parts as the use of the dependency plugin is causing problems

**Comment by calvinjia** [comment]
@jsimsa without the plugin we just lose the ability to enforce the dependencies are properly declared?

**Comment by jsimsa** [comment]
the use of the plugin is problematic because "precise" dependencies means different things for different versions of Hadoop (in particular, the Hadoop 1.x and Hadoop 2.x provide certain classes in in different modules)

I had trouble debugging issue with all the different changes combined into one PR ... so I decided to create this PR as the first milestone

**Comment by jsimsa** [comment]
I might also consider cleaning up the exclusions in this PR

File: tests/pom.xml [hunk 1]
```diff
@@ -108,71 +104,77 @@
...
     <dependency>
       <groupId>org.alluxio</groupId>
       <artifactId>alluxio-core-common</artifactId>
       <version>${project.version}</version>
+      <scope>test</scope>
     </dependency>
     <dependency>
       <groupId>org.alluxio</groupId>
-      <artifactId>alluxio-keyvalue-client</artifactId>
+      <artifactId>alluxio-core-common</artifactId>
```
Comment (yupeng9) [inline tests/pom.xml]: duplicate

File: underfs/s3/pom.xml [hunk 2]
```diff
@@ -30,47 +30,45 @@
...
-    <!-- TODO(jsimsa): Investigate whether the same outcome can be achieved
-     using the promoteTransitiveDependencies option of the maven shade plugin. -->
-    <dependency>
+      <!-- jets3t dependency that needs to included explicitly, because jets3t is shaded. -->
       <groupId>com.jamesmurty.utils</groupId>
       <artifactId>java-xmlbuilder</artifactId>
-      <version>0.4</version>
     </dependency>
     <dependency>
+      <!-- jets3t dependency that needs to included explicitly, because jets3t is shaded. -->
```
Comment (yupeng9) [inline underfs/s3/pom.xml]: why repeating this comment multiple times?

File: assembly/pom.xml [hunk 3]
```diff
@@ -143,7 +144,6 @@
       <plugin>
         <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-shade-plugin</artifactId>
```
Comment (jsimsa) [inline assembly/pom.xml]: it is now managed using plugin management in the top-level module (all versions are)

File: core/client/pom.xml [hunk 4]
```diff
@@ -49,15 +54,20 @@
...
-        <artifactId>alluxio-core-protobuf</artifactId>
-        <version>${project.version}</version>
-      </dependency>
+      <artifactId>alluxio-core-common</artifactId>
+      <version>${project.version}</version>
+    </dependency>
     <dependency>
       <groupId>org.alluxio</groupId>
-      <artifactId>alluxio-core-common</artifactId>
+      <artifactId>alluxio-core-protobuf</artifactId>
```
Comment (jsimsa) [inline core/client/pom.xml]: as per discussion with @calvinjia and @apc999 ... we want to include all direct dependencies instead of relying on transitive ones ... I will have a follow up PR that enforces this

File: core/common/pom.xml [hunk 5]
```diff
@@ -32,19 +32,46 @@
...
+    <dependency>
       <groupId>commons-lang</groupId>
       <artifactId>commons-lang</artifactId>
-      <version>2.6</version>
     </dependency>
     <dependency>
-      <groupId>commons-io</groupId>
-      <artifactId>commons-io</artifactId>
-      <version>2.4</version>
+      <groupId>info.ganglia.gmetric4j</groupId>
```
Comment (jsimsa) [inline core/common/pom.xml]: this has been fixed in branch-1.4 ... let me send you a separate PR for merge branch-1.4 to master

File: core/server/master/pom.xml [hunk 6]
```diff
@@ -33,23 +33,24 @@
   <dependencies>
     <!-- External dependencies -->
     <dependency>
+      <groupId>com.fasterxml.jackson.core</groupId>
```
Comment (jsimsa) [inline core/server/master/pom.xml]: removed

File: core/server/worker/pom.xml [hunk 7]
```diff
@@ -39,23 +39,21 @@
...
     <dependency>
       <groupId>org.alluxio</groupId>
-      <artifactId>alluxio-core-common</artifactId>
+      <artifactId>alluxio-core-client</artifactId>
       <version>${project.version}</version>
     </dependency>
     <dependency>
       <groupId>org.alluxio</groupId>
-      <artifactId>alluxio-core-client</artifactId>
+      <artifactId>alluxio-core-common</artifactId>
```
Comment (jsimsa) [inline core/server/worker/pom.xml]: similar comment as before, we want to include all direct dependencies

File: keyvalue/server/pom.xml [hunk 8]
```diff
@@ -30,34 +30,35 @@
...
+    <!-- Internal dependencies -->
     <dependency>
       <groupId>org.alluxio</groupId>
       <artifactId>alluxio-core-protobuf</artifactId>
       <version>${project.version}</version>
     </dependency>
     <dependency>
       <groupId>org.alluxio</groupId>
-      <artifactId>alluxio-keyvalue-common</artifactId>
+      <artifactId>alluxio-core-server-common</artifactId>
```
Comment (jsimsa) [inline keyvalue/server/pom.xml]: ditto

File: tests/pom.xml (hunk 1 above)
Comment (jsimsa) [inline tests/pom.xml]: fixed

File: underfs/s3/pom.xml (hunk 2 above)
Comment (jsimsa) [inline underfs/s3/pom.xml]: so that dependencies can be moved around or removed without losing this information

File: core/common/pom.xml (hunk 5 above)
Comment (jsimsa) [inline core/common/pom.xml]: https://github.com/Alluxio/alluxio/pull/4785/files

File: core/client/pom.xml (hunk 4 above)
Comment (yupeng9) [inline core/client/pom.xml]: sg

File: core/server/master/pom.xml (hunk 6 above)
Comment (jsimsa) [inline core/server/master/pom.xml]: turns out this is needed for REST API tests to work

File: core/common/pom.xml [hunk 9]
```diff
@@ -32,19 +32,42 @@
...
-      <groupId>commons-io</groupId>
-      <artifactId>commons-io</artifactId>
-      <version>2.4</version>
+      <groupId>io.dropwizard.metrics</groupId>
+      <artifactId>metrics-core</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>io.dropwizard.metrics</groupId>
+      <artifactId>metrics-ganglia</artifactId>
+    </dependency>
```
Comment (calvinjia) [inline core/common/pom.xml]: Is this dependency used?

File: core/server/common/pom.xml [hunk 10]
```diff
@@ -54,42 +54,36 @@
         </exclusion>
       </exclusions>
     </dependency>
+    <!-- Needs to be included before apache-jsp for web UI to work. -->
```
Comment (calvinjia) [inline core/server/common/pom.xml]: Is it possible to remove one? We can do it in another PR, just curious

File: core/common/pom.xml (hunk 9 above)
Comment (jsimsa) [inline core/common/pom.xml]: removed

File: core/server/common/pom.xml (hunk 10 above)
Comment (jsimsa) [inline core/server/common/pom.xml]: I will have a follow up PR to clean this (and exclusions) up

**Comment by calvinjia** [comment]
LGTM
